### PR TITLE
Merchant items

### PR DESCRIPTION
**Pull request by Ashkanthegreat**
- Merchant can add, delete, and edit an item. Merchant cannot edit an item without all required info. Merchant cannot add an item without all required info. 
- Updated Item model  and model test to make it so inventory has to be greater than or equal to 0.
- Updated Item model and model test so price has to be greater than 0.
- Added resources within the merchant namespace, for items.

### CODE REVIEW

File: app/controllers/items_controller.rb [hunk 1]
```diff
@@ -15,15 +15,20 @@ def show
...
+    return @item if params[:item]
+    @item = Hash.new("")
   end
 
   def create
     @merchant = Merchant.find(params[:merchant_id])
     item = @merchant.items.create(item_params)
     if item.save
+      flash[:succss] = "Your item has been saved"
+      redirect_to "/merchant/items" and return if current_merchant?
```
Comment (jpc20) [inline app/controllers/items_controller.rb:27]: We could move this to the merchant items controller.

File: app/controllers/merchant/items_controller.rb [hunk 2]
```diff
@@ -2,4 +2,44 @@ class Merchant::ItemsController < Merchant::BaseController
   def index
     @merchant = Merchant.find(@current_user.merchant_id)
   end
+
+  def update
+    @item = Item.find(params[:id])
```
Comment (Ashkanthegreat) [inline app/controllers/merchant/items_controller.rb:7]: Is there a way to pull out the nested If and else, inside the If?

File: app/controllers/merchant/items_controller.rb [hunk 3]
```diff
@@ -2,4 +2,44 @@ class Merchant::ItemsController < Merchant::BaseController
...
+      flash[:success] = "#{@item.name} is now available for sale"
+      redirect_to "/merchant/items"
+    else
+      @item.update(item_params)
+      if @item.save
+        flash[:succes] = "Item Updated"
+        redirect_to "/merchant/items"
+      else
+        flash[:error] = @item.errors.full_messages.to_sentence
+        render :edit
```
Comment (jpc20) [inline app/controllers/merchant/items_controller.rb:23]: This seems like a long method. Does anyone have any ideas for how to refactor?

File: app/views/merchant/items/edit.html.erb [hunk 4]
```diff
@@ -0,0 +1,22 @@
+<h1>Edit <%=link_to @item.name, "/items/#{@item.id}" %></h1>
+
+<center>
+  <%= form_for [:merchant, @item] do |f|  %>
```
Comment (jpc20) [inline app/views/merchant/items/edit.html.erb:4]: This form routes to the merchant namespace and looks for the items resources.

File: app/controllers/items_controller.rb [hunk 5]
```diff
@@ -15,15 +15,20 @@ def show
...
     @merchant = Merchant.find(params[:merchant_id])
+    return @item if params[:item]
+    @item = Hash.new("")
   end
 
   def create
     @merchant = Merchant.find(params[:merchant_id])
     item = @merchant.items.create(item_params)
     if item.save
+      flash[:succss] = "Your item has been saved"
```
Comment (janegreene) [inline app/controllers/items_controller.rb:26]: Typo to fix here

File: spec/features/users/user_merchant_spec.rb [hunk 6]
```diff
@@ -105,7 +106,200 @@
...
+    fill_in 'Name', with: ""
+    fill_in 'Price', with: 110
+    fill_in 'Description', with: "They're a bit more expensive, and they kinda do pop sometimes, but whatevs.. this is retail."
+    fill_in 'Image', with: ""
+    fill_in 'Inventory', with: -1
+
+    click_button "Update Item"
+
+    expect(page).to have_content("Name can't be blank and Inventory must be greater than or equal to 0")
+    expect(page).to have_button("Update Item")
```
Comment (Lithnotep) [inline spec/features/users/user_merchant_spec.rb:301]: great tests

File: app/controllers/merchant/items_controller.rb [hunk 7]
```diff
@@ -2,4 +2,44 @@ class Merchant::ItemsController < Merchant::BaseController
...
+      flash[:success] = "#{@item.name} is no longer for sale"
+      redirect_to "/merchant/items"
+    elsif params[:type] == "activate"
+      @item.update_attributes(active?: true)
+      flash[:success] = "#{@item.name} is now available for sale"
+      redirect_to "/merchant/items"
+    else
+      @item.update(item_params)
+      if @item.save
+        flash[:succes] = "Item Updated"
```
Comment (janegreene) [inline app/controllers/merchant/items_controller.rb:19]: Typo here for success

File: app/models/item.rb [hunk 8]
```diff
@@ -7,10 +7,10 @@ class Item <ApplicationRecord
   validates_presence_of :name,
                         :description,
                         :price,
-                        :image,
                         :inventory
   validates_inclusion_of :active?, :in => [true, false]
   validates_numericality_of :price, greater_than: 0
+  validates_numericality_of :inventory, greater_than_or_equal_to: 0
```
Comment (janegreene) [inline app/models/item.rb:13]: Wow, thats awesome.

File: app/views/merchant/items/edit.html.erb (hunk 4 above)
Comment (janegreene) [reply app/views/merchant/items/edit.html.erb:4]: Thanks for the explanation, this is great to see.

File: app/controllers/merchant/items_controller.rb (hunk 7 above)
Comment (jpc20) [reply app/controllers/merchant/items_controller.rb]: Thanks!

**Comment by janegreene** [comment]
Form_for for merchant item edit
Merchant item index
merchants can deactivate and reactivate items
Jack resolved comments
tests passing

**Pull request merged**